src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioWorkspaceImpl.AbstractRemoveDocumentUndoUnit.cs

```diff
@@ -0,0 +1,50 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System.Collections.Generic;
+using System.Linq;
+using Microsoft.CodeAnalysis;
+using Microsoft.VisualStudio.OLE.Interop;
+
+namespace Microsoft.VisualStudio.LanguageServices.Implementation.ProjectSystem
+{
+    internal partial class VisualStudioWorkspaceImpl
+    {
+        private abstract class AbstractRemoveDocumentUndoUnit : AbstractAddRemoveUndoUnit
+        {
+            protected readonly DocumentId DocumentId;
+
+            protected AbstractRemoveDocumentUndoUnit(
+                VisualStudioWorkspaceImpl workspace,
+                DocumentId documentId)
+                : base(workspace, documentId.ProjectId)
+            {
+                DocumentId = documentId;
+            }
+
+            protected abstract IReadOnlyList<DocumentId> GetDocumentIds(Project fromProject);
+
+            protected abstract TextDocument GetDocument(Solution currentSolution);
+
+            public override void Do(IOleUndoManager pUndoManager)
+            {
+                var currentSolution = Workspace.CurrentSolution;
+                var fromProject = currentSolution.GetProject(FromProjectId);
+
+                if (fromProject != null &&
+                    GetDocumentIds(fromProject).Contains(DocumentId))
+                {
+                    var updatedProject = fromProject.RemoveDocument(DocumentId);
+                    Workspace.TryApplyChanges(updatedProject.Solution);
```

Comment:
Rather than duplicate this each time, should Do not be abstract as it is, and there's an override for just "here's the project, go transform that"?